src/models/graph.ts

```diff
@@ -372,12 +378,31 @@ export class Graph<N extends INodeBase, E extends IEdgeBase> implements IGraph<N
     return nearestEdge;
   }
 
+  addListener(observer: IObserver): void {
+    this._listeners.push(observer);
+  }
+
+  removeListener(observer: IObserver): void {
+    this._listeners.splice(this._listeners.indexOf(observer), 1);
+  }
+
+  notifyListeners(): void {
+    this._listeners.forEach((listener) => {
+      listener.update();
+    });
+  }
```

Comment:
This code is repeated (in node/edge) which means it would be great to have this outside as a class implementing IObserver and/or ISubject. 

You can also check how nodejs did it with EventEmitter - it is similar pattern.